### PR TITLE
[Triton] Mfma16 support

### PR DESCRIPTION
**Pull request by binarman**
This PR adds support for mfma with tile size 16.

### CODE REVIEW

**Comment by binarman** [comment]
FYI @CaptnJackSparrow

You were interested if we are going to support other types of MFMA instructions in Triton.

File: lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp [hunk 1]
```diff
@@ -152,9 +152,16 @@ class BlockedToMFMA : public mlir::RewritePattern {
     auto ctx = oldAType.getContext();
 
     triton::gpu::MfmaEncodingAttr mfmaEnc;
+
+    int64_t nonKDim = 16;
```
Comment (binarman) [inline lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp:156]: Note: In this PR we emit **only** mfma instructions with size == 16

When #248 is merged, I'll rebase this PR and turn this back to 32.

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp [hunk 2]
```diff
@@ -205,44 +205,50 @@ struct ConvertLayoutOpConversion
     }
 #ifdef USE_ROCM
     if (auto mfmaLayout = layout.dyn_cast<MfmaEncodingAttr>()) {
-      SmallVector<Value> mfmaColIdx(4);
-      SmallVector<Value> mfmaRowIdx(16);
+      unsigned iNonKDim = mfmaLayout.getNonKDim();
+      assert(iNonKDim == 16 || iNonKDim == 32);
```
Comment (weihanmines) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:209]: Would it better to use if the assert statement is replaced by a `if` statement?

Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:209]: Unfortunately, we do not have any fallback path for other values in nonKDim, so I used assert here.

Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:209]: Note for reviewers:
@oplavsic pointed that he wants to refactor this part in his PR: https://github.com/ROCmSoftwarePlatform/triton/pull/247/files#diff-3eee00a71e5ff853d12cbbaaf7971b9d9372080b540b06386acab0e53530e9b9R214

I'll take his implementation

**Comment by binarman** [comment]
So, for GEMM, there are no big difference, but FA suffers for some reason.

Looking into this problem.

Note: that this PR does not enable mfma 16x16, you need to change target mfma type manually in `AccelerateMatmul.cpp`

File: lib/Conversion/TritonGPUToLLVM/ReduceOpToLLVM.cpp [hunk 3]
```diff
@@ -431,10 +431,12 @@ struct ReduceOpConversion
         unsigned shuffleIdx = N;
 #ifdef USE_ROCM
         if (inMfma && inMfma.getIsTransposed()) {
```
Comment (alefimov-amd) [inline lib/Conversion/TritonGPUToLLVM/ReduceOpToLLVM.cpp:433]: @oplavsic 
Could you check if I understood this part correctly.

FA tutorial test passes, but still

**Comment by zhanglx13** [comment]
@binarman Is this one ready for review?

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp [hunk 4]
```diff
@@ -134,8 +136,13 @@ computeTensorElemMapping(ConversionPatternRewriter &rewriter, Location loc,
...
       Value tileHOffset = i32_val(tile * elemsPerInstr[1]);
 
-      Value laneVOffset = urem(laneId, _32);
-      Value laneHOffset = select(icmp_uge(laneId, _32), i32_val(numOfElems), _0);
+      Value laneVOffset = urem(laneId, nonKDim);
+      Value laneHOffset;
+      if (iNonKDim == 32)
+        laneHOffset = select(icmp_uge(laneId, _32), i32_val(numOfElems), _0);
+      else
+        laneHOffset = mul(udiv(laneId, nonKDim), i32_val(numOfElems));
```
[the remark this answers is not in the document]
Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:144]: code with `select` produces slightly llvm IR, which is less coupled and could be scheduled better in some cases.

Unfortunately , now I can not give you an example right now, but it affected performance when we had a problem with high register pressure in FA forward pass.

**Review by zhanglx13 — APPROVED** [review]
LGTM

**Review by jayfurmanek — APPROVED** [review]

**Pull request merged**